---
layout: default
title: Docker
parent: Rules
---

# Docker
{: .no_toc }


## Оглавление
{: .no_toc .text-delta }

1. TOC
{:toc}

---




### Контейнер запускается от имени пользователя root

<span class="d-inline-block p-2 mr-1 v-align-middle bg-red-000"></span>Несоответствующий код:


```java
# Noncompliant code
FROM ubuntu:latest

RUN apt-get update && apt-get install -y \
    software-properties-common \
    python3 \
    python3-pip

COPY . /app
WORKDIR /app

RUN pip3 install -r requirements.txt

CMD ["python3", "app.py"]
```

В этом несоответствующем коде Dockerfile используется для создания образа контейнера для приложения Python. Однако образ основан на образе ubuntu:latest, который включает ненужные пакеты и потенциально подвергает риску безопасность. Кроме того, контейнер запускается от имени пользователя root, что представляет собой угрозу безопасности.



<span class="d-inline-block p-2 mr-1 v-align-middle bg-green-000"></span>Соответствующий код:


```java
# Compliant code
FROM python:3.9-slim

COPY . /app
WORKDIR /app

RUN pip install -r requirements.txt

CMD ["python", "app.py"]

```


В соответствующем коде Dockerfile обновлен для использования базового образа python:3.9-slim, который является облегченным и более безопасным образом, специально разработанным для приложений Python. Это позволяет отказаться от ненужных пакетов и уменьшить поверхность атаки контейнера.

Кроме того, контейнер по умолчанию запускается с пользователем, не являющимся root, что обеспечивает дополнительный уровень безопасности. Инструкции COPY, WORKDIR и RUN соответствующим образом обновлены для работы с новым образом.

Благодаря использованию более безопасного базового образа и запуску контейнера с пользователем, не являющимся пользователем root, соответствующий код снижает риск возникновения уязвимостей и повышает общую безопасность контейнера Docker.

Важно регулярно обновлять базовый образ и зависимости в Dockerfile, чтобы использовать последние исправления безопасности. Кроме того, рассмотрите возможность применения других лучших практик, таких как:

* Использование многоступенчатых сборок для минимизации размера конечного образа и исключения ненужных зависимостей во время сборки.
* Внедрение инструментов сканирования безопасности контейнеров для выявления и устранения уязвимостей в образе.
* Ограничение привилегий и возможностей контейнеров для ограничения потенциальных эксплойтов.
* Использование методов управления секретами для безопасной работы с конфиденциальными данными в контейнере.

Следуя практикам безопасного кодирования и принимая проактивные меры по уменьшению поверхности атаки и устранению уязвимостей, вы можете повысить безопасность и устойчивость своих контейнеров Docker.
Чтобы решить эту проблему, приведем пример соответствующего кода:



